pragma-node/src/infra/repositories/onchain_repository.rs

```diff
@@ -251,3 +254,217 @@ pub async fn get_checkpoints(
+            publisher IN ('{publishers_list}')
+        GROUP BY 
+            publisher;
+        "#,
+        table_name = table_name,
+        publishers_list = publishers_list,
+    );
+
+    let conn = pool.get().await.map_err(adapt_infra_error)?;
+    let updates = conn
+        .interact(move |conn| diesel::sql_query(raw_sql).load::<RawPublisherUpdates>(conn))
+        .await
+        .map_err(adapt_infra_error)?
+        .map_err(adapt_infra_error)?;
+
+    Ok(updates)
+}
+
+async fn get_publisher_with_components(
+    pool: &Pool,
+    table_name: &str,
+    publisher: &RawPublisher,
+    publisher_updates: &RawPublisherUpdates,
+    currencies: &HashMap<String, BigDecimal>,
+) -> Result<Publisher, InfraError> {
+    let raw_sql_entries = format!(
+        r#"
+        SELECT
+            entries.pair_id,
+            entries.price,
+            entries.source,
+            entries.timestamp as last_updated_timestamp
+        FROM
+            {table_name} entries
+        INNER JOIN (
+            SELECT
+                pair_id,
+                MAX(timestamp) AS max_timestamp
+            FROM
+                {table_name}
+            WHERE
+                publisher = '{publisher_name}'
+            GROUP BY
+                pair_id
+        ) AS latest ON entries.pair_id = latest.pair_id AND entries.timestamp = latest.max_timestamp
+        WHERE
+            entries.publisher = '{publisher_name}'
+        ORDER BY
+            entries.pair_id, entries.source ASC;
+        "#,
+        table_name = table_name,
+        publisher_name = publisher.name
+    );
+
+    let conn = pool.get().await.map_err(adapt_infra_error)?;
+
+    let raw_components = conn
+        .interact(move |conn| {
+            diesel::sql_query(raw_sql_entries).load::<RawLastPublisherEntryForPair>(conn)
+        })
+        .await
+        .map_err(adapt_infra_error)?
+        .map_err(adapt_infra_error)?;
+
+    let components: Vec<PublisherEntry> = raw_components
+        .into_iter()
+        .map(|component| component.to_publisher_entry(currencies))
+        .collect();
+
+    let last_updated_timestamp = components
+        .iter()
+        .map(|component| component.last_updated_timestamp)
+        .max()
+        .unwrap_or_default();
+
+    let publisher = Publisher {
+        publisher: publisher.name.clone(),
+        website_url: publisher.website_url.clone(),
+        last_updated_timestamp,
+        r#type: publisher.publisher_type as u32,
+        nb_feeds: publisher_updates.nb_feeds as u32,
+        daily_updates: publisher_updates.daily_updates as u32,
+        total_updates: publisher_updates.total_updates as u32,
+        components,
+    };
+    Ok(publisher)
+}
+
+pub async fn get_publishers_with_components(
+    pool: &Pool,
+    publishers: Vec<RawPublisher>,
+    network: Network,
+    currencies: HashMap<String, BigDecimal>,
+) -> Result<Vec<Publisher>, InfraError> {
+    let table_name = get_table_name(network, DataType::SpotEntry);
```

Comment:
let's add a query param `data_type` then

Reply:
moved to pragma_common & added!